File: packages/sdk/examples/03-bulk-ingest.ts
/**
 * 03-bulk-ingest.ts — Bulk-ingest the cookbook corpus.
 *
 * Demonstrates: bulkIngestOrchestrate(), progress callback,
 * on_existing policy. Re-runs are cheap because of
 * skip_if_unchanged.
 *
 * Env: same as 01-quick-start.ts.
 *
 * Run:
 *     npx tsx 03-bulk-ingest.ts
 *
 * See also:
 *     SDK_COOKBOOK_OUTLINE.md §4
 */

import { TextralClient, bulkIngestOrchestrate } from '@textral/sdk';
import { readFile } from 'node:fs/promises';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));

const profileName = process.env.TEXTRAL_PROFILE;
const namespace = process.env.TEXTRAL_NAMESPACE ?? 'cookbook';
const providerKeyRef = process.env.TEXTRAL_PROVIDER_KEY_REF ?? 'openai';

const client = profileName
  ? new TextralClient({ profile: profileName })
  : new TextralClient({
      baseUrl: process.env.TEXTRAL_BASE_URL ?? 'https://api.textral.alacrity.ai',
      apiKey: process.env.TEXTRAL_API_KEY ?? '',
    });

const filenames = ['alexandria.md', 'lighthouse.md', 'scholars.md', 'decline.md'];
const files = await Promise.all(
  filenames.map(async (name) => {
    const buf = await readFile(join(__dirname, 'data', name));
    return {
      filename: name,
      bytes: new Uint8Array(buf.buffer, buf.byteOffset, buf.byteLength),
      size_bytes: buf.byteLength,
      content_type: 'text/markdown',
    };
  }),
);

const result = await bulkIngestOrchestrate(client, {
  namespace,
  config: {
    embedding: {
      provider: 'openai',
      model: 'text-embedding-3-large',
      dimensions: 1536,
      provider_key_ref: providerKeyRef,
    },
    chunking: { profile: 'generic', target_tokens: 600, overlap_tokens: 80 },
    mode: 'full',
  },
  files,
  on_existing: 'skip_if_unchanged',
  auto_finalize: true,
  concurrency: 4,
  onProgress: (snap) => {
    console.log(
      `${snap.state.padEnd(12)} ${snap.counts.succeeded}/${snap.total_files} (${snap.progress_pct}%)`,
    );
  },
});

console.log('---');
console.log(`bulk_job_id: ${result.bulk_job_id}`);
console.log(`final state: ${result.final_status.state}`);
console.log(
  `succeeded: ${result.final_status.counts.succeeded}, ` +
    `skipped: ${result.final_status.counts.skipped}, ` +
    `failed: ${result.final_status.counts.failed}`,
);

File: packages/sdk-python/examples/03_bulk_ingest.py
"""03_bulk_ingest.py — Bulk-ingest the cookbook corpus.

Demonstrates: bulk_ingest_orchestrate(), progress callback,
on_existing policy. Re-runs are cheap because of
skip_if_unchanged.

Env: same as 01_quick_start.py.

Run:
    python 03_bulk_ingest.py

See also:
    docs/development/sdks/SDK_COOKBOOK_OUTLINE.md §4
"""

from __future__ import annotations

import os
from pathlib import Path

from textral import (
    BulkOrchestrateFile,
    Client,
    bulk_ingest_orchestrate,
)

DATA_DIR = Path(__file__).parent / "data"
PROVIDER_KEY_REF = os.environ.get("TEXTRAL_PROVIDER_KEY_REF", "openai")
NAMESPACE = os.environ.get("TEXTRAL_NAMESPACE", "cookbook")


def main() -> None:
    filenames = ["alexandria.md", "lighthouse.md", "scholars.md", "decline.md"]
    files = [
        BulkOrchestrateFile(
            filename=name,
            bytes_=(DATA_DIR / name).read_bytes(),
            size_bytes=(DATA_DIR / name).stat().st_size,
            content_type="text/markdown",
        )
        for name in filenames
    ]

    def on_progress(snap: dict) -> None:
        state = str(snap.get("state", "?")).ljust(12)
        counts = snap.get("counts", {})
        total = snap.get("total_files", 0)
        succeeded = counts.get("succeeded", 0)
        pct = snap.get("progress_pct", 0)
        print(f"{state} {succeeded}/{total} ({pct}%)")

    with Client() as client:
        result = bulk_ingest_orchestrate(
            client,
            namespace=NAMESPACE,
            config={
                "embedding": {
                    "provider": "openai",
                    "model": "text-embedding-3-large",
                    "dimensions": 1536,
                    "provider_key_ref": PROVIDER_KEY_REF,
                },
                "chunking": {"profile": "generic", "target_tokens": 600, "overlap_tokens": 80},
                "mode": "full",
            },
            files=files,
            on_existing="skip_if_unchanged",
            auto_finalize=True,
            concurrency=4,
            on_progress=on_progress,
        )

    counts = result.final_status.get("counts", {})
    print("---")
    print(f"bulk_job_id: {result.bulk_job_id}")
    print(f"final state: {result.final_status.get('state')}")
    print(
        f"succeeded: {counts.get('succeeded', 0)}, "
        f"skipped: {counts.get('skipped', 0)}, "
        f"failed: {counts.get('failed', 0)}"
    )


if __name__ == "__main__":
    main()
